contracts/vesting/src/vesting.rs

```diff
@@ -0,0 +1,559 @@
+                        self.get_delegated_vesting(block_time, env, storage)
+                            .amount
+                            .u128(),
+                    ),
+            ),
+            denom: DENOM.to_string(),
+        }
+    }
+
+    fn spendable_coins(
+        &self,
+        block_time: Option<Timestamp>,
+        env: &Env,
+        storage: &dyn Storage,
+    ) -> Coin {
+        Coin {
+            amount: Uint128(
+                self.get_balance(storage)
+                    .u128()
+                    .saturating_sub(self.locked_coins(block_time, env, storage).amount.u128()),
+            ),
+            denom: DENOM.to_string(),
+        }
+    }
+
+    fn get_vested_coins(&self, block_time: Option<Timestamp>, env: &Env) -> Coin {
+        let block_time = block_time.unwrap_or(env.block.time);
+        let period = self.get_current_vesting_period(block_time);
+
+        match period {
+            // We're in the first period, or the vesting has not started yet.
+            0 => Coin {
+                amount: Uint128(0),
+                denom: DENOM.to_string(),
+            },
+            1..=7 => Coin {
+                amount: Uint128(self.tokens_per_period() * period as u128),
+                denom: DENOM.to_string(),
+            },
+            _ => Coin {
+                amount: self.coin.amount,
+                denom: DENOM.to_string(),
+            },
+        }
+    }
+
+    fn get_vesting_coins(&self, block_time: Option<Timestamp>, env: &Env) -> Coin {
+        Coin {
+            amount: Uint128(
+                self.get_original_vesting().amount.u128()
+                    - self.get_vested_coins(block_time, env).amount.u128(),
+            ),
+            denom: DENOM.to_string(),
+        }
+    }
+
+    fn get_start_time(&self) -> Timestamp {
+        self.start_time
+    }
+
+    fn get_end_time(&self) -> Timestamp {
+        self.periods[(NUM_VESTING_PERIODS - 1) as usize].end_time()
+    }
+
+    fn get_original_vesting(&self) -> Coin {
+        self.coin.clone()
+    }
+
+    fn get_delegated_free(
+        &self,
+        block_time: Option<Timestamp>,
+        env: &Env,
+        storage: &dyn Storage,
+    ) -> Coin {
+        self.get_delegated_with_op(&lt, block_time, env, storage)
+    }
+
+    fn get_delegated_vesting(
+        &self,
+        block_time: Option<Timestamp>,
+        env: &Env,
+        storage: &dyn Storage,
+    ) -> Coin {
+        self.get_delegated_with_op(&ge, block_time, env, storage)
+    }
+}
+
+fn lt(x: u64, y: u64) -> bool {
+    x < y
+}
+
+fn ge(x: u64, y: u64) -> bool {
+    x >= y
+}
+
+pub fn populate_vesting_periods(start_time: u64, n: u64) -> Vec<VestingPeriod> {
+    let mut periods = Vec::new();
+    // There are eight 3 month periods in two years
+    for i in 0..n {
+        let period = VestingPeriod {
+            start_time: start_time + i * VESTING_PERIOD,
+        };
+        periods.push(period);
+    }
+    periods.shrink_to_fit();
```

Comment:
Rather than having to shrink to fit, why not create the `periods` with appropriate capacity from the start, i.e. `Vec::with_capacity(n as usize)`? Alternatively create the vector from Iterator

Reply:
Good call